### PR TITLE
Fix SSO on workers

### LINKED ISSUE #8966: workers.md gives incorrect advice about various endpoints
https://github.com/matrix-org/synapse/blob/develop/docs/workers.md#synapseappgeneric_worker says:

> Additionally, the following endpoints should be included if Synapse is configured to use SSO...
> ```
> # OpenID Connect requests.
> ^/_matrix/client/(api/v1|r0|unstable)/login/sso/redirect$
> ^/_synapse/oidc/callback$
> 
> # SAML requests.
> ^/_matrix/client/(api/v1|r0|unstable)/login/sso/redirect$
> ^/_matrix/saml2/authn_response$
> 
> # CAS requests.
> ^/_matrix/client/(api/v1|r0|unstable)/login/(cas|sso)/redirect$
> ^/_matrix/client/(api/v1|r0|unstable)/login/cas/ticket$
> ```

However, none of these endpoints are actually enabled on worker processes, so any attempts to send them to a worker will result in a 404.

### PR DESCRIPTION
**Pull request by richvdh**
Fixes #8966.

~~Based on #9262.~~ Commits should be reviewable independently.

### CODE REVIEW

File: synapse/storage/databases/main/registration.py [hunk 1]
```diff
@@ -443,6 +443,26 @@ def f(txn):
 
         return await self.db_pool.runInteraction("get_users_by_id_case_insensitive", f)
 
+    async def record_user_external_id(
```
Comment (richvdh) [inline synapse/storage/databases/main/registration.py:446]: this is moving from `RegistrationStore` to `RegistrationWorkerStore`. There's no caching on this table so I think it's safe to move.

File: docs/workers.md [hunk 2]
```diff
@@ -225,7 +225,6 @@ expressions:
     ^/_matrix/client/(api/v1|r0|unstable)/joined_groups$
     ^/_matrix/client/(api/v1|r0|unstable)/publicised_groups$
     ^/_matrix/client/(api/v1|r0|unstable)/publicised_groups/
-    ^/_synapse/client/password_reset/email/submit_token$
```
Comment (richvdh) [inline docs/workers.md:228]: tangential, but we may as well fix it while we're here. 

This line was added in #8227. AFAICT, it has never worked on a worker.

File: synapse/rest/synapse/client/__init__.py [hunk 3]
```diff
@@ -42,6 +42,18 @@ def build_synapse_client_resource_tree(hs: "HomeServer") -> Mapping[str, Resourc
...
+    # rely on optional dependencies.
+    if hs.config.oidc_enabled:
+        from synapse.rest.oidc import OIDCResource
+
+        resources["/_synapse/oidc"] = OIDCResource(hs)
+
+    if hs.config.saml2_enabled:
+        from synapse.rest.saml2 import SAML2Resource
+
+        resources["/_matrix/saml2"] = SAML2Resource(hs)
```
Comment (clokep) [inline synapse/rest/synapse/client/__init__.py:55]: Does it make sense to note that this is under `_matrix` for backwards compat?

File: docs/workers.md [hunk 4]
```diff
@@ -256,25 +256,28 @@ Additionally, the following endpoints should be included if Synapse is configure
...
     # OpenID Connect requests.
-    ^/_matrix/client/(api/v1|r0|unstable)/login/sso/redirect$
     ^/_synapse/oidc/callback$
 
     # SAML requests.
-    ^/_matrix/client/(api/v1|r0|unstable)/login/sso/redirect$
     ^/_matrix/saml2/authn_response$
 
     # CAS requests.
-    ^/_matrix/client/(api/v1|r0|unstable)/login/(cas|sso)/redirect$
```
Comment (clokep) [inline docs/workers.md:268]: Note that this one was slightly different in that it has `(cas|sso)` in it. Not sure if any clients still use the `cas` path though.

File: docs/workers.md [hunk 5]
```diff
@@ -256,25 +256,28 @@ Additionally, the following endpoints should be included if Synapse is configure
...
     ^/_matrix/client/(api/v1|r0|unstable)/login/cas/ticket$
 
-Note that a HTTP listener with `client` and `federation` resources must be
-configured in the `worker_listeners` option in the worker config.
-
-Ensure that all SSO logins go to a single process (usually the main process). 
+Ensure that all SSO logins go to a single process.
```
Comment (clokep) [inline docs/workers.md:273]: This is probably true for the UI auth endpoints too?

**Review by clokep — APPROVED** [review]
Looks good generally! I left a couple of comments, but nothing major.

File: synapse/rest/synapse/client/__init__.py (hunk 3 above)
Comment (richvdh) [reply synapse/rest/synapse/client/__init__.py:55]: sorry, note where? as a comment here?

File: docs/workers.md (hunk 4 above)
Comment (richvdh) [reply docs/workers.md:268]: yeahh, I'm hoping that's old enough that basically nobody will encounter it.

File: docs/workers.md (hunk 5 above)
Comment (richvdh) [reply docs/workers.md:273]: I'm not sure it is? afaik nothing is tracked in-memory for UIA (at least for CAS/OIDC; SAML has https://github.com/matrix-org/synapse/issues/7530)

Comment (clokep) [reply docs/workers.md:273]: Right, the username mapping sessions and extra attributes only matter on login, not during UI auth. OK.

File: synapse/rest/synapse/client/__init__.py (hunk 3 above)
Comment (clokep) [reply synapse/rest/synapse/client/__init__.py:55]: As a comment, yes. (My thinking being that this is in the `synapse.rest.synapse` package someone might be confused in the future at why we're registering a `_matrix` URL here. 🤷 ) Not a big deal either way though.

Comment (richvdh) [reply synapse/rest/synapse/client/__init__.py:56]: ok, I added a comment.

**Pull request merged**